jdrfandroidbleparser/src/main/java/org/ehealthinnovation/jdrfandroidbleparser/bgm/characteristic/compoundcharacteristic/GlucoseMeasurementCharacteristic.kt

```diff
@@ -0,0 +1,141 @@
+                        unit = Units.MOLE_PER_LITRE
+                    else
+                        unit = Units.KILOGRAM_PER_LITRE
+                }
+                if (it.contains(Flags.STATUS_ANNUNCIATION_PRESENT))
+                    annunciationFlags = SensorStatusAnnunciation.parseFlags(getNextIntValue(c, BluetoothGattCharacteristic.FORMAT_UINT16))
+            }
+
+            //format the date, time, offset into java type
+            displayTime = BluetoothDateTime.getDate(timeYear, timeMonth, timeDay, timeHour, timeMinute, timeSecond, timeOffset)
+            errorFreeParse = true
+        } catch (e: NullPointerException) {
+            Log.e(tag, nullValueException)
+        }
+        return errorFreeParse
+    }
+
+    /**
+     * Get the glucose measurement concentration. It provide unit conversion as well.
+     *
+     * @param outputUnit the expected output concentration unit.
+     * @return the value returned.
+     */
+    fun getConcentration(outputUnit: Units): Float? {
+        var outputValue: Float? = -1.0f
+        val CONVERSION_FACTOR_KG_PER_LITRE_TO_MMOL_PER_LITRE = 5555.5556f
+        val CONVERSION_FACTOR_MOL_PER_LITRE_TO_MMOL_PER_LITRE = 18000f
+        val CONVERSION_FACTOR_KG_PER_LITRE_TO_MG_PER_DECILITRE = 100000f
+        val CONVERSION_FACTOR_MOL_PER_LITRE_TO_MG_PER_DECILITRE = 18000f
+        val CONVERSION_FACTOR_MOL_PER_LITRE_TO_KG_PER_LITRE = 0.18f
+        val CONVERSION_FACTOR_KG_PER_LITRE_TO_MOL_PER_LITRE = 5.5556f
+
+        if (outputUnit === unit) {
+            return concentration
+        }
+
+        when (outputUnit) {
+            Units.MILLIMOLE_PER_LITRE -> {
+                if (unit === Units.KILOGRAM_PER_LITRE) {
+                    outputValue = CONVERSION_FACTOR_KG_PER_LITRE_TO_MMOL_PER_LITRE * concentration!!
+                } else if (unit === Units.MOLE_PER_LITRE) {
+                    outputValue = CONVERSION_FACTOR_MOL_PER_LITRE_TO_MMOL_PER_LITRE * concentration!!
+                }
+            }
+            Units.MILLIGRAM_PER_DECILITRE -> {
+                if (unit === Units.KILOGRAM_PER_LITRE) {
+                    outputValue = CONVERSION_FACTOR_KG_PER_LITRE_TO_MG_PER_DECILITRE * concentration!!
+                } else if (unit === Units.MOLE_PER_LITRE) {
+                    outputValue = CONVERSION_FACTOR_MOL_PER_LITRE_TO_MG_PER_DECILITRE * concentration!!
+                }
+            }
+            Units.KILOGRAM_PER_LITRE -> {
+                if (unit === Units.MOLE_PER_LITRE) {
+                    outputValue = CONVERSION_FACTOR_MOL_PER_LITRE_TO_KG_PER_LITRE * concentration!!
+                }
+            }
+            Units.MOLE_PER_LITRE -> if (unit === Units.KILOGRAM_PER_LITRE) {
+                outputValue = CONVERSION_FACTOR_KG_PER_LITRE_TO_MOL_PER_LITRE * concentration!!
+            }
+            else -> throw IllegalArgumentException("Unsupport output unit" + "$outputUnit")
+        }
+        return outputValue
+    }
```

Comment:
Why is this included in the data class? Is the class supposed to do the conversion for us? Or is this a legacy piece of code for assisting UI population? If it's not used directly in the data class to satisfy the needs of the spec parsing, it should be removed.

Reply:
It is for the convenient of the user. The this class as data class will be passed around to UI and database processing methods. By then, methods consuming this class can retrieve the measurement value in local unit by calling getConcentration. This is necessary unless we want to shift the burden of unit conversion elsewhere.

Reply:
Hey I'm just looking at this code review again, this code needs to get taken out of here before we can merge.